File: src/cpp/arc.cpp
#include "arc.h"

#include "glgraphics.h"

Arc::Arc(Permutation *permutation, Slot *slotU, Slot *slotV)
	:m_permutation(permutation),
	 m_slotU(slotU),
	 m_slotV(slotV),
	 m_startTime(SDL_GetTicks())
{
}

void Arc::computeParameters()
{
}

void Arc::adjustCenter(const vec2f &)
{
}

void Arc::normalize(float)
{
}

void Arc::render() const
{
	GLGraphics *glGraphics = GLGraphics::getInstance();
	glGraphics->setColor(m_permutation->getColor());
	
	Uint32 curTime = SDL_GetTicks();
	float duration = (float)(curTime - m_startTime) / sc_animDuration;
	float t0 = duration - static_cast<int>(duration);

	unsigned int numDots = 3;
	for (unsigned int i = 0; i < numDots; ++i)
	{
		float t = t0 + static_cast<float>(i) / numDots;
		if (t > 1.0) t -= 1.0;
		vec2f p = getPoint(t);
		glGraphics->drawCircle(p, 0.01, true);
	}
}



LineArc::LineArc(Permutation *permutation, Slot *slotU, Slot *slotV)
	:Arc(permutation, slotU, slotV)
{
}

vec2f LineArc::getPoint(float t) const
{
	return (1 - t) * m_slotU->getCenter() + t * m_slotV->getCenter();
}

void LineArc::render() const
{
	Arc::render();
	
	GLGraphics *glGraphics = GLGraphics::getInstance();
	glGraphics->setColor(m_permutation->getColor());

	vec2f v0 = m_slotU->getCenter();
	vec2f v1 = m_slotV->getCenter();

	glGraphics->drawLine(v0, v1);
}



CircleArc::CircleArc(Permutation *permutation,
                     Slot *slotU,
                     Slot *slotV,
                     float circleRadius,
                     bool circlePlus,
                     bool circleInverted)
	:Arc(permutation, slotU, slotV),
	 m_circleRadius(circleRadius),
	 m_circlePlus(circlePlus),
	 m_circleInverted(circleInverted),
	 m_circleCenter(0.0, 0.0),
	 m_thetaU(0.0),
	 m_thetaV(0.0),
	 m_dTheta(0.0)
{
	computeParameters();
}

void CircleArc::computeParameters()
{
	vec2f v0 = m_slotU->getCenter();
	vec2f v1 = m_slotV->getCenter();

	// Calculate the center of the circle
	float d = (v0 - v1).getNorm();
	float r2 = m_circleRadius * m_circleRadius;
	float disc = r2 - d * d / 4.0;
	float minDisc = 0.001;
	if (disc < minDisc) disc = minDisc;
	float h = sqrt(disc);
	vec2f midpoint = (v0 + v1) / 2.0;
	vec2f disp = (v1 - v0).getUnit();
	vec2f rot = vec2f(disp.getY(), -disp.getX());
	if (m_circlePlus) rot *= -1.0;
	m_circleCenter = midpoint + h * rot;

	// Set up interval of angles to iterate along
	vec2f deltaU = v0 - m_circleCenter;
	vec2f deltaV = v1 - m_circleCenter;

    m_thetaU = atan2(deltaU[1], deltaU[0]);
	m_thetaV = atan2(deltaV[1], deltaV[0]);

	m_dTheta = m_thetaV - m_thetaU;

	// Put dTheta into the interval [0, 2pi]
	while (m_dTheta < 0)             m_dTheta += 2.0 * 3.14159;
	while (m_dTheta > 2.0 * 3.14159) m_dTheta -= 2.0 * 3.14159;

	// Invert it?
	if (m_circleInverted)
	{
		float tmp = m_thetaU;
		m_thetaU = m_thetaV;
		m_thetaV = tmp;
		m_dTheta = 2.0 * 3.14159 - m_dTheta;
	}
}

vec2f CircleArc::getPoint(float t) const
{
	//float theta = m_thetaU + t * m_dTheta;
	float theta = m_thetaU + t * m_dTheta;

	vec2f p(cos(theta), sin(theta));
	p = m_circleRadius * p + m_circleCenter;

	return p;
}

void CircleArc::render() const
{
	Arc::render();
	
	GLGraphics *glGraphics = GLGraphics::getInstance();
	glGraphics->setColor(m_permutation->getColor());

	unsigned int numIntervals = 20;
	if (sc_debugEnabled) numIntervals = 5;
	float dt = 1.0 / numIntervals;
	for (unsigned int i = 0; i < numIntervals; ++i)
	{
		float t0 = i * dt;
		float t1 = (i + 1) * dt;

		vec2f p0 = getPoint(t0);
		vec2f p1 = getPoint(t1);

		glGraphics->drawLine(p0, p1);

		if (sc_debugEnabled)
		{
			glGraphics->setColor(t0, 0.0, 1.0 - t0);
			glGraphics->drawCircle(p0, 0.03, true);
			glGraphics->setColor(m_permutation->getColor());
		}
	}

	if (sc_debugEnabled)
	{
		glGraphics->setColor(0.0, 1.0, 0.0);
		glGraphics->drawCircle(m_circleCenter, 0.03, true);
	}
}

void CircleArc::normalize(float radius)
{
	m_circleRadius /= radius;
}

File: src/js/arc.js
var ARC_LINE = 0;
var ARC_CIRCLE = 1;

function NewLineArc(permutation, slotU, slotV)
{
	return new Arc(permutation, slotU, slotV, ARC_LINE);
}

function NewCircleArc(permutation, slotU, slotV, circleRadius, circlePlus, circleInverted)
{
	var ret = new Arc(permutation, slotU, slotV, ARC_CIRCLE);
	
	ret.circleRadius = circleRadius;
	ret.circlePlus = circlePlus;
	ret.circleInverted = circleInverted;

	ret.computeParameters();

	ret.debugEnabled = true;

	return ret;
}

class Arc
{
	constructor(permutation, slotU, slotV, shape)
	{
		this.permutation = permutation;
		this.slotU = slotU;
		this.slotV = slotV;
		this.shape = shape;
		this.startTime = 0;
		this.debugEnabled = false;

		// CircleArc vars
		this.circleRadius = 0.0;
		this.circlePlus = false;
		this.circleInverted = false;
		this.circleCenter = new vec2f(0.0, 0.0);
		this.thetaU = 0.0;
		this.thetaV = 0.0;
		this.dTheta = 0.0;
	}

	draw(context)
	{
		var color = this.permutation.getColor();
		
		if (this.shape == ARC_LINE)
		{
			var v0 = this.slotU.getCenter();
			var v1 = this.slotV.getCenter();

			var color = this.permutation.getColor();

			drawLine(context, color, v0.x, v0.y, v1.x, v1.y);
		}
		else if (this.shape == ARC_CIRCLE)
		{
			var numIntervals = 20;
			if (this.debugEnabled) numIntervals = 5;
			var dt = 1.0 / numIntervals;
			
			for (var i = 0; i < numIntervals; i++)
			{
				var t0 = i * dt;
				var t1 = (i + 1) * dt;

				var p0 = this.getPoint(t0);
				var p1 = this.getPoint(t1);

				drawLine(context, color, p0.x, p0.y, p1.x, p1.y);

				if (this.debugEnabled)
				{
					var r = Math.floor(255 * t0).toString(16);
					while (r.length < 2)
						r = '0' + r;
					var g = '00';
					var b = Math.floor(255 * (1.0 - t0)).toString(16);
					while (b.length < 2)
						b = '0' + b;
					var purple = '#' + r + g + b;
					fillCircle(context, purple, p0.x, p0.y, 3);
				}
			}

			if (this.debugEnabled)
			{
				var green = '#00FF00';
				var x = this.circleCenter.x;
				var y = this.circleCenter.y;
				fillCircle(context, green, x, y, 3);
			}
		}
	}

	getPoint(t)
	{
		var ret = new vec2f(0.0, 0.0);
		
		if (this.shape == ARC_LINE)
		{
			var lhs = this.slotU.getCenter().scale(1.0 - t);
			var rhs = this.slotV.getCenter().scale(t);
			ret = lhs.add(rhs);
		}
		else if (this.shape == ARC_CIRCLE)
		{
			var theta = this.thetaU + t * this.dTheta;

			var p = new vec2f(Math.cos(theta), Math.sin(theta));
			p = p.scale(this.circleRadius).add(this.circleCenter);

			return p;
		}

		return ret;
	}

	computeParameters()
	{
		if (this.shape == ARC_CIRCLE)
		{
			var v0 = this.slotU.getCenter();
			var v1 = this.slotV.getCenter();

			// Calculate the center of the circle;
			var d = v0.sub(v1).norm();
			var r2 = this.circleRadius * this.circleRadius;
			var disc = r2 - d * d / 4.0;
			var minDisc = EPSILON;
			if (disc < minDisc) disc = minDisc;
			var h = Math.sqrt(disc);
			var midpoint = v0.add(v1).scale(0.5);
			var disp = v1.sub(v0).unit();
			var rot = new vec2f(disp.y, -disp.x);
			if (this.circlePlus) rot = rot.scale(-1.0);
			this.circleCenter = midpoint.add(rot.scale(h));

			// Set up interval of angles to iterate along
			var deltaU = v0.sub(this.circleCenter);
			var deltaV = v1.sub(this.circleCenter);

			this.thetaU = Math.atan2(deltaU.y, deltaU.x);
			this.thetaV = Math.atan2(deltaV.y, deltaV.x);

			this.dTheta = this.thetaV - this.thetaU;

			// Put dTheta into the interval [0, 2pi]
			while (this.dTheta < 0)             this.dTheta += 2.0 * Math.PI;
			while (this.dTheta > 2.0 * Math.PI) this.dTheta -= 2.0 * Math.PI;

			// Invert it?
			if (this.circleInverted)
			{
				var tmp = this.thetaU;
				this.thetaU = this.thetaV;
				this.thetaV = tmp;
				this.dTheta = 2.0 * Math.PI - this.dTheta;
			}
		}
	}

	adjustCenter(vec)
	{
	}

	normalize(radius)
	{
		if (this.shape == ARC_CIRCLE)
		{
			this.circleRadius /= radius;
		}
	}
}
